### ISSUE TITLE
Sub X509 certificate mode: allow X509 to have multiple TLS-Crypt-V2 keys

### ISSUE THREAD

**Issue by TinCanTech** [TinCanTech/easy-tls#104]
Easy-TLS _can_ act as a fine-grain control of a single client X509 cert for multiple client nodes.

Something along the lines of:
- `--sub-cert-node=A3 build-tls-crypt-v2-client s01 c01 tct ab0123456789` 
This would create a TLS Crypt V2 key called: `c01-A3-tls-crypt-v2.key`

**Comment by TinCanTech** [comment]
For those of you who understand such things, this is **not** commonly known as "_a key array_"

**Comment by TinCanTech** [comment]
This is more tricky than it looked at first - `inline_index_serial_to_common_name` will return multiple records because the X509 Client cert is the serial number to key on and that number is used for all sub-cert keys.

**Comment by TinCanTech** [comment]
Not so bad after all ;-)

**Comment by TinCanTech** [comment]
@houmie this would allow you have multiple TLS CV2 keys per x509 Cert .. interested ?

**Comment by houmie** [comment]
Hi @TinCanTech 
That sounds exciting. How does this affect the overall outcome? 
1) This could mean that several OpenVPN servers should share the same X.509 certificate. Correct? If correct, it would allow more flexibility. Do you have any real life examples of how this could be beneficial? 
2) Could this have any effect on hiding the footprint even better from deep packet inspections? (As the current version already does)
3) Is there any other advantage I couldn't think of?

**Comment by TinCanTech** [comment]
The main benefit which I can see is that you can have multiple VPN profiles all using different TLS CV2 Keys but all using the same x509 cert.  So if one key gets matched by DPI another may not ;-)

**Comment by houmie** [comment]
Ah yes, makes sense. Sounds intriguing. :-) I'm happy to put some time aside and test this next week, if this is ready. Is there any documentation how this is setup compared to the classic way?

**Comment by TinCanTech** [comment]
There is no documentation, other than this issue.

Essentially, you would do this:
* Complete your EasyRSA PKI, with all your required certs.
* Use `easytls` to create Server TLS Crypt V2 key and Inline Server details.
* Use `easytls` to create Client TLS crypt V2 keys and _sub-certificate_ keys.
  Note: These are not subordinate keys in any way but that is the word I chose for now.
* Use `easytls` to create Inline files for your each of your Client keys and _sub-certificate_ keys.

As a Client you would have something like:
* houmie x509 cert x1, key x1
* houmie TLS Crypt V2 key x1 - Inlined with X509 above.
* houmie TLS Crypt V2 sub-key 1 - Inlined with X509 above.
* houmie TLS Crypt V2 sub-key 2 - Inlined with X509 above.
* etc ..

Travellers _may_ find they can use different keys if they are blocked by DPI.

There would also be details for other clients of yours and they can all have sub-keys.

**Comment by TinCanTech** [comment]
#110

**Comment by houmie** [comment]
Awesome.  I will get to this 100% and test it. I'm away this week but return on Sunday.  I definitely get it done.  👍🏼

**Comment by TinCanTech** [comment]
I am in the process of reviewing the new code for this. It is quite a large change but overall I am quite pleased with it.  In the process of writing this I had to extend my test suite and also really drill down into the lower functions to ensure the correct data is returned. It can still be improved but it is getting sweeter ;-)

**Comment by TinCanTech** [comment]
#64

**Comment by TinCanTech** [comment]
Please let me know if you experience any problems. Please post your logs.

**Comment by houmie** [comment]
Hello. Happy new year, mate.
Yes, I should be back by Sunday (day after tomorrow) and can get started. I keep you in the loop. Thanks

**Comment by houmie** [comment]
Alrighty, as promised I'm back.

I have setup a fresh server with OpenVPN 2.5.0 installed.

I thought I double check the steps with you.

```
git clone https://github.com/TinCanTech/easy-tls
cp ./easy-tls/easytls /etc/openvpn/easy-rsa/
cd /etc/openvpn/easy-rsa/
./easytls init-tls
cd /etc/openvpn/easy-rsa/pki/issued/
SERVER=$(basename server*.crt .crt)
cd ~
CLIENT=$(basename client*.ovpn .ovpn)
cd /etc/openvpn/easy-rsa
./easytls --openvpn=/usr/sbin/openvpn build-tls-crypt-v2-server $SERVER
```

In the old way I would continue creating the client like this:
```
./easytls --openvpn=/usr/sbin/openvpn build-tls-crypt-v2-client $SERVER $CLIENT
```

But I think now instead of that, I should be creating sub-certificate keys straight away?

```./easytls --help``` doesn't really make it clear how to do it.

May you provide more details, please?

**Comment by TinCanTech** [comment]
You can create your first key in the same way and Inline it as before.

You can also use `--custom-group=houmie` above and below.  If you specify a custom group then use it in all your commands, those which need it will use it.

To create a sub-key use:
`./easytls --sub-key-name=$Your_choice --custom-group=houmie --openvpn=/usr/sbin/openvpn build-tls-crypt-v2-client $SERVER $CLIENT`

The second key would be called:
`./easytls/${CLIENT}-${Your_choise}-tls-crypt-v2.key`


To Inline the second key use:
`./easytls --sub-key-name=$Your_choice --openvpn=/usr/sbin/openvpn inline-tls-crypt-v2 $CLIENT`

The Inline file will be called:
`./easytls/${CLIENT}-${Your_choise}.inline`

**Comment by houmie** [comment]
Alrighty,

So I have now executed the following and am logging them here so you can see the progress:

```
./easytls --custom-group=houmie --openvpn=/usr/sbin/openvpn build-tls-crypt-v2-server $SERVER

TLS crypt v2 server key created: /etc/openvpn/easy-rsa/pki/easytls/server_jtjdMPkGoxv9w8PT-tls-crypt-v2.key

```

And now client:

```
./easytls --custom-group=houmie --sub-key-name=mySubKey --openvpn=/usr/sbin/openvpn build-tls-crypt-v2-client $SERVER $CLIENT

Custom-Group field for metadata: houmie

Type the word 'yes' to continue, or any other input to abort.
  Is the Custom-Group field correct ? yes

Sub-key-name field for metadata: mySubKey

Type the word 'yes' to continue, or any other input to abort.
  Is the Sub-key-name correct ? yes

TLS crypt v2 client key created: /etc/openvpn/easy-rsa/pki/easytls/clientDeTestM2-mySubKey-tls-crypt-v2.key
```

Very interesting, as I never have used custom groups before. If I have let say three servers in three different locations. They should ideally share the same custom group, I take it?

As see above, the second key is generated successfully.

However the next call fails:

```
./easytls --custom-group=houmie --sub-key-name=mySubKey --openvpn=/usr/sbin/openvpn inline-tls-crypt-v2 $CLIENT

Inline TLS crypt v2 file created: /etc/openvpn/easy-rsa/pki/easytls/clientDeTestM2-mySubKey.inline

Easy-TLS version: 1.27

Easy-TLS error:

Not found: inline_file = /etc/openvpn/easy-rsa/pki/easytls/clientDeTestM2.inline
```

The subway is not attached to the path, hence it fails finding the file.

**Comment by TinCanTech** [comment]
Can you please try the last again like so:
`./easytls inline-remove $CLIENT`
`./easytls --verbose --custom-group=houmie --sub-key-name=mySubKey --openvpn=/usr/sbin/openvpn inline-tls-crypt-v2 $CLIENT
`
I think the issue is coming from the auto-check.

Also,
The custom group can be thought of as a general identification marker.
The way I use it is, all TLS-Crypt-V2 keys which I generate for myself have `tincantech` in the field.  But it could equally have some sub-structure, eg: `tincantech-branch-01`. It is up to you how you want to use it.  During development, I tried to foresee requirements.

**Comment by TinCanTech** [comment]
Also, please pull `git/master/easy-tls` first.

**Comment by houmie** [comment]
Sure.

```
./easytls inline-remove $CLIENT

Easy-TLS version: 1.27

Easy-TLS error:



Missing: /etc/openvpn/easy-rsa/pki/easytls/clientDeTestM2.inline
```

I have pulled the latest before ```cp ./easy-tls/easytls /etc/openvpn/easy-rsa/```

I can have taken an image just before starting with tls.  I can easily roll it back, if you wish. But it seems ```inline-remove``` isn't working correctly.

**Comment by TinCanTech** [comment]
OK .. I think I can see the issue but I'm not sure how to fix it yet.

I believe that you created a Sub-key before creating the default key and Inline file.
That should work fine but for some reason the inline hash check is failing to set the correct name for the file.  I think I can replicate this and then work on a fix.  Thanks for testing, I'll let you know how I get on.

**Comment by TinCanTech** [comment]
Ah, another thought though .. try this:
`./easytls --verbose --custom-group=houmie --sub-key-name=mySubKey inline-remove $CLIENT`
then
`./easytls --verbose --custom-group=houmie --sub-key-name=mySubKey --openvpn=/usr/sbin/openvpn inline-tls-crypt-v2 $CLIENT`

**Comment by TinCanTech** [comment]
OK, I know what the problem is:

`inline_index_common_name_list ()` used by `status_easyrsa_renewed ()` does not pick up Sub-key names.

**Comment by TinCanTech** [comment]
I have what appears to be a fix, if you find time please test it:

https://github.com/TinCanTech/easy-tls/blob/e4dc44379f16b153b990820211202224482aa9a1/easytls#L759
```
	#etls_valid_name_list="$(inline_index_common_name_list)"
	etls_valid_name_list="$(inline_index_common_name_sub_name_list)"
```
Replace line 759 with the code above and retest.  Thanks for your help and no need to rush ;-)

**Comment by houmie** [comment]
No problem. Happy to help. 👍🏼 

So I changed the line:
```
# Print known inline files from Easy-TLS
status_easytls_inline ()
{
        [ -f "$EASYTLS_INLINE_INDEX" ] || return 0
        #etls_valid_name_list="$(inline_index_common_name_list)"
        etls_valid_name_list="$(inline_index_common_name_sub_name_list)"
```

This still throws

```
./easytls inline-remove $CLIENT

Easy-TLS version: 1.27

Easy-TLS error:

Missing: /etc/openvpn/easy-rsa/pki/easytls/clientDeTestM2.inline
```

But this worked:

```
root@ip-172-31-36-140:/etc/openvpn/easy-rsa# ./easytls --verbose --custom-group=houmie --sub-key-name=mySubKey inline-remove $CLIENT
Note: using Easy-RSA configuration from: ./vars

Remove: /etc/openvpn/easy-rsa/pki/easytls/clientDeTestM2-mySubKey.inline

Type the word 'yes' to continue, or any other input to abort.
  Remove inline file ? yes
Inline-index hash check OK
  DEL: clientDeTestM2[[:blank:]]0946A3F16D93B1A0E93DDB665C1F10BD[[:blank:]]bf6314bf9b188dbca21ef546971d54f07a8dd3d5[[:blank:]]mySubKey
Inline Index Update complete!

Inline file removed: /etc/openvpn/easy-rsa/pki/easytls/clientDeTestM2-mySubKey.inline

* AUTO-CHECK
* status invalid
* status renewed
* Verify inline index hash
Inline-index hash check OK
* Verify disabled list hash
Disabled-list hash check OK

```

What do you think?

**Comment by TinCanTech** [comment]
That looks correct to me :+1: 

The reason the first command fails is because you do not have an Inline file named:
`clientDeTestM2.inline`

Your file **is** called:
`clientDeTestM2-mySubKey.inline`

And so, with the change it seems to be working correctly for you now.

**Comment by houmie** [comment]
Great. So here is the next one:

```
./easytls --verbose --custom-group=houmie --sub-key-name=mySubKey --openvpn=/usr/sbin/openvpn inline-tls-crypt-v2 $CLIENT
Note: using Easy-RSA configuration from: ./vars
Inline-index hash check OK
  ADD: clientDeTestM2 0946A3F16D93B1A0E93DDB665C1F10BD bf6314bf9b188dbca21ef546971d54f07a8dd3d5 mySubKey
Inline Index Update complete!

Inline TLS crypt v2 file created: /etc/openvpn/easy-rsa/pki/easytls/clientDeTestM2-mySubKey.inline

* AUTO-CHECK
* status invalid
* status renewed
* Verify inline index hash
Inline-index hash check OK
* Verify disabled list hash
Disabled-list hash check OK
```

**Comment by TinCanTech** [comment]
You successfully removed and then recreated your inline file.

LGTM

**Comment by houmie** [comment]
Excellent. So in the classic way I would copy the crypt-v2.key into client.ovpn like this:

```
sed -i '/tls-crypt/,$d' $MYHOME/$CLIENT.ovpn
echo "<tls-crypt-v2>" >> $MYHOME/$CLIENT.ovpn
cat /etc/openvpn/easy-rsa/pki/easytls/client*-tls-crypt-v2.key >> $MYHOME/$CLIENT.ovpn
echo "</tls-crypt-v2>" >> $MYHOME/$CLIENT.ovpn
```

How should I proceed with this new inline file that is generated?
```
/etc/openvpn/easy-rsa/pki/easytls/clientDeTestM2-mySubKey.inline
```

**Comment by TinCanTech** [comment]
Changing TLS Keys on your client and not your server is a recipe for disaster.

No. What you do with Easy-TLS Inline files is this:
* Copy/paste the Inline file directly into your OpenVPN config file.
Or 
* Use OpenVPN `--config houmie-key1.inline` in your config file.
Where the Inline file becomes a parameter of your main OpenVPN server/client config file.

**Comment by TinCanTech** [comment]
I do not intend for Easy-TLS to make complete config files for OpenVPN.

**Comment by houmie** [comment]
Good morning, Yes that's fine, I'm happy to create the config files for OpenVPN by myself. I just need to understand it better so I can automate and test them on my end.

1) So that I understand this correctly. OpenVPN is injecting by default ```<tls-crypt>``` into the client.ovpn. So you recommend not to replace that with ```<tls-crypt-v2>```, but instead insert only the inline into the server config file? 

2) Do I then just remove ```<tls-crypt>...</tls-crypt>``` entirely from the client config files in this case?

3) Can a server config list multiple inlines from different clients? How is the <tag> defined that needs to go into server config? Is it still ```<tls-crypt-v2> Content of Inline key </tls-crypt-v2>``` or something else?

Thanks

**Comment by TinCanTech** [comment]
Whatever you choose to use on the server, you must also use on the client.

`--tls-auth` and `--tls-crypt` are simple pre-shared keys. The same key must be used at both ends. `--tls-auth` also requires `--key-direction`: Server uses `0`, client uses `1`.

`--tls-crypt-v2` uses a server specific key and a client specific key. These are not a _pre-shared key_. The client key can only be used the connect to the server key which was used to create it.

At both ends, instead of specifying `--ca`, `--cert`, `--key` and `--tls-*`, you can simply use `--config /etc/openvpn/easy-rsa/pki/easytls/vpn-node-01.inline`.

Automating the use of the Inline files seems like over-kill to me.

**Comment by houmie** [comment]
Good evening mate,

I see.  Of course, sorry I should have looked into the inline file to see it for myself. It simply includes all four below, which no longer will be required in the server config.

```
tls-crypt-v2 tls-crypt-v2.key 0
key server_vSpTuNo38WJBclg2.key
cert server_vSpTuNo38WJBclg2.crt
ca ca.crt
```

So far so good. This will not change anything with client.ovpn, correct?  In there everything stays the same way.

Now to the next test.  You mentioned that multiple VPN profiles can now all be using different TLS CV2 Keys but all using the same x509 cert. How can I create a second client ovpn that can do that?

That means it will have the same ```verify-x509-name server_vSpTuNo38WJBclg2 name```, same ```ca```, same ```cert``` and same ```key```, different ```tls-crypt-v2``` and still be able to connect to the server?

or could you explain the use case, please ?

**Comment by TinCanTech** [comment]
> I see. Of course, sorry I should have looked into the inline file to see it for myself. It simply includes all four below, which no longer will be required in the server config.
> 
> ```
> tls-crypt-v2 tls-crypt-v2.key 0
> key server_vSpTuNo38WJBclg2.key
> cert server_vSpTuNo38WJBclg2.crt
> ca ca.crt
> ```
> 
> So far so good. 

That sounds correct. (Except, you do not require the direction parameter `0` or `1` for any `--tls-crypt*` key)

> This will not change anything with client.ovpn, correct? In there everything stays the same way.

Not sure I understand.  Your client must use the correct credentials ..

**Comment by TinCanTech** [comment]
> That means it will have the same `verify-x509-name server_vSpTuNo38WJBclg2 name`, same `ca`, same `cert` and same `key`, different `tls-crypt-v2` and still be able to connect to the server?

Yes, that is the idea.

> or could you explain the use case, please ?

Sure, 

The server side is fairly straight forward, you only have the one `--tls-crypt-v2` server key.
The client is similar if you only have one `--tls-crypt-v2` key.

The client is slightly different when using the `--sub-key-name` option.

Think of it like this:
```
* X509 CA =
          |
          |- X509 Server01 Cert/Key = 
          |                         |- TLS-Crypt-V2 Server key - Inline file Server01
          |
          |- X509 Client01 Cert/key = 
          |                         |- TLS-Crypt-V2 Client key 01 - Inline file Client01-subkey01
          |                         |- TLS-Crypt-V2 Client key 02 - Inline file Client01-subkey02
          |

```

**Comment by TinCanTech** [comment]
In order to create multiple subkeys do this:
```
./easytls --sub-key-name=test01 --custom-group=tincantech build-tls-crypt-v2-client client01 abdcef012345 aabbccddeeff
./easytls --sub-key-name=test02 --custom-group=tincantech build-tls-crypt-v2-client client01 abdcef012345 aabbccddeeff
./easytls --sub-key-name=test01 inline-tls-crypt-v2 client01
./easytls --sub-key-name=test02 inline-tls-crypt-v2 client01
```

**Comment by TinCanTech** [comment]
I first named this option `--sub-cert-key=NAME` but decided that `--sub-key-name=NAME` was simpler.  However, I am not convinced this name is the right one yet..  Suggestions welcome.

**Comment by houmie** [comment]
Ah very nice. It is now coming together in my head. Allow me to sleep over this and I'll continue tomorrow. 

I think ```--sub-key-name``` is the better name.  Because ```cert``` and ```key``` are usually describing two different things and it would get confusing if you went with ```--sub-cert-key```.  So you chose the better name already, I would say.

**Comment by TinCanTech** [comment]
Updated help: https://github.com/TinCanTech/easy-tls/wiki/15-Multiple-TLS-Crypt-V2-Keys-per-X509-Certificate

**Comment by TinCanTech** [comment]
> Ah very nice. It is now coming together in my head. Allow me to sleep over this and I'll continue tomorrow.

Many thanks for taking time to help me.
 
> I think `--sub-key-name` is the better name. Because `cert` and `key` are usually describing two different things and it would get confusing if you went with `--sub-cert-key`. So you chose the better name already, I would say.

I agree because of confusing cert and key in one option. 

But _technically_, **it is** a key which is used in a _sub certificate_ fashion ;-)
One cert -> many keys .. I could call it `--icing-sugar` :candy:

**Comment by houmie** [comment]
> Many thanks for taking time to help me.

No problem. My pleasure. 
 
> But _technically_, **it is** a key which is used in a _sub certificate_ fashion ;-)
> One cert -> many keys .. I could call it `--icing-sugar` 🍬

ha. :-)

The documentation that you just added made it super clear.  So I could have one server and three clients for different devices, such as iPhone, Laptop and Android phone.  If one of the devices get caught by DPI in future, the other devices will still work.

Another scenario is that I could write a native phone app and integrate OpenVPN in it. Then I could include three sub-keys in the same app.  The user could then switch between the sub keys on the same phone.  So when one of the sub-keys is caught, switching to another sub-key will make it work again. Correct?

**Comment by TinCanTech** [comment]
That is the idea.. but beware, it _may_ work, **not** it _will_ work again.

This is the problem with DPI; _they_ have all the power and no matter what we do, it is still possible that this trick can be blocked. But it is quite a nice trick on our side, for a change.

**Comment by houmie** [comment]
Hey buddy,

Sorry I'm a bit struggling with time.  I will finish the testing by this weekend.  

My plan is to create two sub keys for two separate clients and connect to the same server to see if it all works smoothly.
Thanks

**Comment by TinCanTech** [comment]
There is no rush. You have already helped me a lot and I am grateful for your time.

**Comment by TinCanTech** [comment]
I would only ask that you make sure to use `git/master` for your test. :candy:

**Comment by TinCanTech** [comment]
@houmie I cannot thank you enough for the help and effort **you** have already made.

I have only one further request, please use option `--verbose` all the time, in future.

:cake:

**Comment by TinCanTech** [comment]
I am thinking, this command could have a `x10 keys` function .. ?
Automatically create 10 keys and inline files.

**Comment by houmie** [comment]
Hi buddy,

I've been testing everything on master via verbose flag. Everything looks solid.  But the VPN connection test is still outstanding. I hope to get that done tomorrow morning my time.

I have one question:

If I create two sub-key-clients:

```
clientDeTestM2-mySubKey1-tls-crypt-v2.key
clientDeTestM2-mySubKey2-tls-crypt-v2.key
```

The server would recognise and accept either one of them.
The subKeys above could be used, each on a separate device with a separate OPENVPN client.

Or

the same device could import two OpenVPN profiles one with mySubKey1 included and the other profile with mySubKey2.
That way the same device could switch between them every other day to change the foot print and therefore hide better from DPI.

Have I understood it correctly?
Thanks

**Comment by TinCanTech** [comment]
You have understood correctly.  Both of your methods above are valid.

All TLS-Cryp-V2 Client keys, regardless of `--sub-key-name`, can be used by the Server key which created them. (Just be careful if you start using _hardware-lockdown_.. )

**Comment by houmie** [comment]
> I am thinking, this command could have a `x10 keys` function .. ?
> Automatically create 10 keys and inline files.

Yes, I think this would be useful. 👍🏼

**Comment by TinCanTech** [comment]
Closing in favour of interactive menus
